ros2topic/ros2topic/verb/hz.py

```diff
@@ -86,20 +87,70 @@ def main(self, *, args):
-            return eval_fn
-        filter_expr = expr_eval(args.filter_expr)
-    else:
-        filter_expr = None
+def _get_nested_messages(msg_class):
+    all_attributes = list(msg_class.__slots__)
+    for attr in msg_class.__slots__:
+        value = getattr(msg_class, attr)
+        if hasattr(value, '__slots__'):
+            nested_messages = _get_nested_messages(value)
+            all_attributes.extend(nested_messages)
+    return all_attributes
+
+def _setup_base_safe_eval():
+    safe_eval_model = base_eval_model.clone()
+
+    # extend base_eval_model
+    safe_eval_model.nodes.extend(['Call', 'Attribute', 'List', 'Tuple', 'Dict', 'Set',
+                                  'ListComp', 'DictComp', 'SetComp', 'comprehension',
+                                  'Mult', 'Pow', 'boolop', 'mod', 'Invert',
+                                  'Is', 'IsNot', 'FloorDiv', 'If', 'For'])
+
+    # allow-list safe Python built-in functions
+    safe_builtins = [
+        'abs', 'all', 'any', 'bin', 'bool', 'chr', 'cmp', 'divmod', 'enumerate',
+        'float', 'format', 'hex', 'id', 'int', 'isinstance', 'issubclass',
+        'len', 'list', 'long', 'max', 'min', 'ord', 'pow', 'range', 'reversed',
+        'round', 'slice', 'sorted', 'str', 'sum', 'tuple', 'type', 'unichr',
+        'unicode', 'xrange', 'zip', 'filter', 'dict', 'set', 'next'
+    ]
+
+    safe_eval_model.allowed_functions.extend(safe_builtins)
+    return safe_eval_model
+
+def _setup_safe_eval(safe_eval_model, msg_class, topic):
+    # allow-list topic builtins, msg attributes
+    topic_builtins = [i for i in dir(topic) if not i.startswith('_')]
+    safe_eval_model.attributes.extend(topic_builtins)
+    # recursively get all nested message attributes
+    msg_attributes = _get_nested_messages(msg_class)
+    safe_eval_model.attributes.extend(msg_attributes)
+    return safe_eval_model
+
 
+def main(args):
     with DirectNode(args) as node:
-        _rostopic_hz(node.node, topic, window_size=args.window_size, filter_expr=filter_expr,
-                     use_wtime=args.use_wtime)
+        topics = args.topic_name
+        filter_expr = None
+        # set up custom safe eval model for filter expression
+        if args.filter_expr:
+            safe_eval_model = _setup_base_safe_eval()
+            for topic in topics:
+                msg_class = get_msg_class(
+                    node, topic, blocking=True, include_hidden_topics=True)
+                if msg_class is None:
+                    continue
+
+                safe_eval_model = _setup_safe_eval(safe_eval_model, msg_class, topic)
+
+            def expr_eval(expr):
+                def eval_fn(m):
+                    safe_expression = Expr(expr, model=safe_eval_model)
+                    return eval(safe_expression.code)
+                return eval_fn
+            filter_expr = expr_eval(args.filter_expr)
+
+        _rostopic_hz(node.node, topics, qos_args=args, window_size=args.window_size,
```

Comment:
Also, `_rostopic_hz()` here does not expect the `qos_args` arg, see https://github.com/ros2/ros2cli/issues/1002.